### ISSUE TITLE
Rewrite links to AMP version

### ISSUE THREAD

**Issue by janrieke** [ampproject/amphtml#2663]
As a user, I would expect that links (`<a href="...">`) in an AMP document also go to the AMP version of the linked document (if available). For internal links, this can easily be implemented, because we know for which documents AMP versions exist.
However, it is unfeasible for external links: We would need to parse the linked document's HTML, check if it is AMP, or if not, try to find a `<link rel="amphtml">` tag. Furthermore, an AMP version could possibly appear or disappear later.
Therefore, it should be the task of the CDN to automatically rewrite links if an AMP version exists. Additionally, there should be a flag in AMP to deactivate this link rewriting if it is not desired by the publisher.

**Comment by stephanfowler** [comment]
Very interesting proposal. I imagine some publishers wouldn't want this, seeing AMPs purely as stepping-stones to their canonical site. But other publishers might be motivated by it to start experimenting with AMP homepages or fully AMP'd sites. So, because of the likely differing appetites for it, the deactivation option seems crucial.

**Comment by kevinsinger** [comment]
Good proposal - Could be an interesting experiment.
However, I would prefer the link rewriting to be deactivated by default and have an activation option instead -  So the publishers have more control about when they want to test this option.

**Comment by fletc3her** [comment]
I think the automatic rewriting of URLs would work well for us, helping to ensure that users stayed on AMP pages when possible, and continuing to serve traffic from the CDN.

I've implemented links to AMP content from within the AMP pages on my sites. If the user arrives at the site through AMP they will continue to be served AMP pages. There is an option in the footer which allows them to return to the Desktop Site similar to how many mobile sites work. However, those links go to the local AMP pages directly, not through the CDN. I have considered changing all the links to point to the CDN, but I'm still chasing down validation errors and don't want users to hit the invalid AMP page error that the CDN returns.

**Comment by davidwhthomas** [comment]
Just noting, it is also possible to get the list of AMP URLs for link rewrite using the [AMP URL API](https://developers.google.com/amp/cache/reference/acceleratedmobilepageurl/rest/)

**Comment by fletc3her** [comment]
On sites I manage the AMP pages aren't performing as well as the original pages in terms of user engagement and ad revenue so we definitely want to shunt users over to our regular site as quickly as possible. In the future we might prefer to keep users who start on an AMP URLs on AMP pages, but it wouldn't be beneficial for us to have the links rewritten automatically.

**Comment by drdamour** [comment]
@fletc3her i agree, rewriting <a> tags wouldn't be right...but rewriting <amp-link> tags would be cool as it implies the intent

**Comment by aonurdemir** [comment]
I am not sure that the discussion is addressing my problem. Therefore, I would like to describe my problem:

I have a blog having a lot of articles. Also have amp pages(links) of them. In each article, there are links to other articles, which are also amp links.(Those links are created in the server side if an amp version exists. If not, non-amp links are placed.)
When a user views one of my articles in the google cache by a link like "https://cdn.ampproject.org/c/s/<url_here>", if s/he clicks to navigate to another article, s/he is redirected to the amp link sitting on my original domain. I want user to navigate in google cache without being redirected to my domain by finding a way to replace the links pointing to my domain with the ones pointing to google cache urls when the page is viewed in google cache.

**Comment by drdamour** [comment]
Y close?

**Comment by Gregable** [comment]
Sorry, I should have provided some reasoning. This is not really an issue with AMP, it's a Google Cache request. There is currently no plan for Google Cache to implement this.
